walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,134 @@
+        occupiedTimeSlots.remove(timeSlot);
+        addTimeRange(occupiedTimeSlots, addedTimeSlot);
+        return occupiedTimeSlots;
+      }
+      // New TimeRange overlaps with an already existing meeting.
+      if (timeSlot.overlaps(addedTimeSlot)) {
+        addTimeRange(occupiedTimeSlots, mergeTimeRanges(timeSlot, addedTimeSlot));
+        return occupiedTimeSlots;
+      }
+    }
+    // No special-casing applied to TimeRange added normally.
+    occupiedTimeSlots.add(addedTimeSlot);
+    return occupiedTimeSlots;
+  }
+
+  /*
+   * Merges overlapping TimeRanges into a single TimeRange that includes both of them.
+   */
+  private TimeRange mergeTimeRanges(TimeRange a, TimeRange b) {
+    return TimeRange.fromStartEnd(a.start() < b.start() ? a.start() : b.start(),
+        a.end() > b.end() ? a.end() : b.end(), /*inclusive=*/false);
+  }
+
+  /*
+   * Returns a collection with TimeRanges that are not available for meetings.
+   */
+  private ArrayList<TimeRange> getOccupiedTimeSlots(Collection<Event> events,
+      HashSet<String> requestAttendees) {
+    ArrayList<TimeRange> occupiedTimeRangesForRequestAttendees = new ArrayList<TimeRange>();
+    for (Event meeting : events) {
+      for (String attendee : meeting.getAttendees()) {
+        // If the meeting contains a requested attendee it is added.
+        if (requestAttendees.contains(attendee)) {
+          occupiedTimeRangesForRequestAttendees = addTimeRange(occupiedTimeRangesForRequestAttendees,
+              meeting.getWhen());
+          break;
+        }
+      }
+    }
+    Collections.sort(occupiedTimeRangesForRequestAttendees, TimeRange.ORDER_BY_START);
+    return occupiedTimeRangesForRequestAttendees;
+  }
+
+  /*
+   * Calculates whether requested meeting's duration fits in between occupied TimeRanges.
+   */
+  private Collection<TimeRange> getFreeTimeSlots(ArrayList<TimeRange> occupiedTimeSlots, 
+      long duration) {
+    Collection<TimeRange> possibleTimeSlots = new ArrayList<>();
+    // Return all available if there are no occupied TimeRanges.
+    if (occupiedTimeSlots.isEmpty()) {
+      possibleTimeSlots.add(TimeRange.WHOLE_DAY);
+      return possibleTimeSlots;
+    }
+    // From midnight to the first meeting.
+    if (occupiedTimeSlots.get(0).start() - TimeRange.START_OF_DAY >= duration) {
+      possibleTimeSlots.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, occupiedTimeSlots.get(0).start(), false));
+    }
+    int previousEnd = occupiedTimeSlots.get(0).end();
+    // Check for spaces between occupied TimeRanges.
+    for (int iterator = 1; iterator < occupiedTimeSlots.size(); iterator++) {
+      int currentStart = occupiedTimeSlots.get(iterator).start();
+      int diff = currentStart - previousEnd;
+      if (currentStart - previousEnd >= duration) {
+        possibleTimeSlots.add(TimeRange.fromStartEnd(previousEnd, currentStart, false));
+      }
+      previousEnd = occupiedTimeSlots.get(iterator).end();
+    }
```

Comment:
I think this section could use a little more clarification, whether it's comments or utilizing helper functions.

Reply:
Wrote comments and erased unnecessary code.